File: data_structures/queues/min_priority_queue.cc
#include "min_priority_queue.h"

int main() {
  std::cout << "Test min priority queue: " << std::endl;
  int arr[10] = {4, 1, 3, 2, 16, 9, 10, 14, 8, 7};
  data_structures::queues::MinPriorityQueue<int> min_priority_queue(arr);
  std::cout << min_priority_queue;
  min_priority_queue.DecreaseKey(8, 6);
  std::cout << min_priority_queue;

  return 0;
}

File: java/src/main/java/org/knguyenc/datastructures/queues/MinPriorityQueue.java
package org.knguyenc.datastructures.queues;

import org.knguyenc.datastructures.heaps.MinHeap;

public class MinPriorityQueue<T extends Comparable<T>> extends MinHeap<T> {

    public MinPriorityQueue() {
        super();
    }

    public MinPriorityQueue(int capacity) {
        super(capacity);
    }

    public T minimum() {
        return data[0];
    }

    public T extractMin() {
        if (isEmpty()) throw new Error("Priority queue underflow, can not extract max");
        return remove(0);
    }

    public void decreaseKey(int index, T key) {
        if (index < 0 && index >= size) throw new Error("Invalid index");
        if (key.compareTo(data[index]) > 0) throw new Error("New key is larger than current key");
        data[index] = key;
        while (index > 0 && data[index].compareTo(data[parent(index)]) < 0) {
            swap(index, parent(index));
            index = parent(index);
        }
    }

    @Override
    public void insert(T value) {
        if (isFull()) throw new Error("Priority queue overflow, can not insert");
        data[size++] = value;
        decreaseKey(size - 1, value);
    }
}
